src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -825,6 +825,17 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
                             | "cttz_nonzero"
                             | "ctlz"
                             | "ctlz_nonzero" => is_const_fn = Some(def_id),
+                            "transmute" => {
+                                is_const_fn = Some(def_id);
```

Comment:
this needs to happen inside the `mode != Fn`, otherwise we are promoting `transmute` (treating anything as const inside `Fn` makes it immediately promotable)

still needs to be outside the feature gate check in order to have the better diagnostic